lib/kino_explorer/data_frame_cell.ex

```diff
@@ -0,0 +1,332 @@
+      missing_require: ctx.assigns.missing_require
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_info({:scan_binding_result, data_options, data_frame_alias, missing_require}, ctx) do
+    ctx =
+      assign(ctx,
+        data_options: data_options,
+        data_frame_alias: data_frame_alias,
+        missing_require: missing_require
+      )
+
+    updated_fields =
+      case {ctx.assigns.root_fields["data_frame"], data_options} do
+        {nil, [%{variable: data_frame} | _]} -> updates_for_data_frame(data_frame)
+        _ -> %{}
+      end
+
+    ctx = if updated_fields == %{}, do: ctx, else: assign(ctx, updated_fields)
+
+    broadcast_event(ctx, "set_available_data", %{
+      "data_options" => data_options,
+      "fields" => updated_fields,
+      "require" => missing_require
+    })
+
+    {:noreply, ctx}
+  end
+
+  @impl true
+  def handle_event("update_field", %{"field" => "data_frame", "value" => value}, ctx) do
+    updated_fields = updates_for_data_frame(value)
+    ctx = assign(ctx, updated_fields)
+    broadcast_event(ctx, "update_data_frame", %{"fields" => updated_fields})
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_field", %{"operation" => nil, "field" => field, "value" => value}, ctx) do
+    parsed_value = parse_value(field, value)
+    ctx = update(ctx, :root_fields, &Map.put(&1, field, parsed_value))
+    broadcast_event(ctx, "update_root", %{"fields" => %{field => parsed_value}})
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_field", %{"operation" => "filters", "field" => "column"} = fields, ctx) do
+    {column, idx} = {fields["value"], fields["idx"]}
+    updated_filter = updates_for_filters(column, ctx)
+    updated_operation = List.replace_at(ctx.assigns.operations["filters"], idx, updated_filter)
+    updated_operations = %{ctx.assigns.operations | "filters" => updated_operation}
+    ctx = assign(ctx, operations: updated_operations)
+
+    broadcast_event(ctx, "update_operation", %{
+      "operation" => "filters",
+      "idx" => idx,
+      "fields" => updated_filter
+    })
+
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_field", %{"operation" => operation} = fields, ctx) do
+    {field, value, idx} = {fields["field"], fields["value"], fields["idx"]}
+    parsed_value = parse_value(field, value)
+
+    updated_operation =
+      put_in(ctx.assigns.operations[operation], [Access.at(idx), field], parsed_value)
+
+    updated_operations = %{ctx.assigns.operations | operation => updated_operation}
+    ctx = assign(ctx, operations: updated_operations)
+
+    broadcast_event(ctx, "update_operation", %{
+      "operation" => operation,
+      "idx" => idx,
+      "fields" => %{field => parsed_value}
+    })
+
+    {:noreply, ctx}
+  end
+
+  def handle_event("add_operation", %{"operation" => operation}, ctx) do
+    new_operation = String.to_atom(operation) |> default_operation()
+    updated_operation = ctx.assigns.operations[operation] ++ [new_operation]
```

Comment:
Maybe let's distinguish `operation_type` or similar, otherwise it's hard to tell if it's an atom or a map :)